cargo-screeps/src/config.rs

```diff
@@ -0,0 +1,178 @@
+    pub mode: DeployMode,
+    pub copy: Option<CopyConfiguration>,
+    pub upload: Option<UploadConfiguration>,
+}
+
+#[derive(Clone, Debug)]
+pub struct UploadConfiguration {
+    pub username: String,
+    pub password: String,
+    pub hostname: String,
+    pub ssl: bool,
+    pub port: i32,
+    pub ptr: bool,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct CopyConfiguration {
+    pub dest: PathBuf,
+    #[serde(default)]
+    pub prune: bool,
+}
+
+#[derive(Fail, Debug)]
+pub enum ConfigError {
+    #[fail(display = "missing username")]
+    MissingUsername,
+    #[fail(display = "missing password")]
+    MissingPassword,
+}
+
+impl TryFrom<FileUploadConfiguration> for UploadConfiguration {
+    type Error = ConfigError;
+
+    fn try_from(value: FileUploadConfiguration) -> Result<UploadConfiguration, Self::Error> {
+        let FileUploadConfiguration {
+            username,
+            password,
+            hostname,
+            ssl,
+            port,
+            ptr,
+        } = value;
+
+        let hostname = hostname.unwrap_or_else(|| "screeps.com".into());
+        let ssl = ssl.unwrap_or_else(|| hostname == "screeps.com");
+        let port = port.unwrap_or_else(|| if ssl { 443 } else { 80 });
+        let username = username.ok_or(ConfigError::MissingUsername)?;
+        let password = password.ok_or(ConfigError::MissingPassword)?;
+
+        Ok(UploadConfiguration {
+            username,
+            password,
+            hostname,
+            ssl,
+            port,
+            ptr,
+        })
+    }
+}
+
+impl TryFrom<FileConfiguration> for Configuration {
+    type Error = ConfigError;
+
+    fn try_from(value: FileConfiguration) -> Result<Configuration, Self::Error> {
+        let FileConfiguration {
+            mode,
+            branch,
+            old_upload,
+            upload,
+            copy,
+            output_wasm_file,
+            output_js_file,
+        } = value;
+
+        let upload = Some(upload.unwrap_or(old_upload).try_into()?);
+
+        let branch = branch.unwrap_or_else(|| "default".into());
+        let output_js_file = output_js_file.unwrap_or_else(|| "main.js".into());
+        let output_wasm_file = output_wasm_file.unwrap_or_else(|| "compiled.wasm".into());
+
+        Ok(Configuration {
+            branch,
+            mode,
+            upload,
+            copy,
+            output_js_file,
+            output_wasm_file,
+        })
+    }
+}
+
+impl Configuration {
+    pub fn read<P: AsRef<Path>>(config_file: P) -> Result<Self, failure::Error> {
+        let config_file = config_file.as_ref();
+        ensure!(
+            config_file.exists(),
+            "expected configuration to exist at {}",
+            config_file.display(),
+        );
+
+        let config_str = {
+            use std::io::Read;
+            let mut buf = String::new();
+            fs::File::open(config_file)
+                .context("opening config file")?
```

Comment:
Oh right - one other thing. Failure's `.context()` is not going to be too useful, since the new error message does not actually include the original one - it just replaces whatever the error was with "opening config file" without any mention of the OS error.

I'd prefer to use specific error variants instead if that's alright. We could also change the handler in `main` to print out all error causes, that would also work.